### DISCUSSION TITLE
Extension: pg-repack

### DISCUSSION THREAD

**Discussion by isdinAghmir** [cloudnative-pg/cloudnative-pg#1632] [Ideas]
Hello, 

I've been using the operator for a few weeks now and i am very satisfied. 

I am also wondering if you plan to add something that can let us install extensions such as pg-repack (which is a life saver for removing bloat from tables when you can only afford minimal downtime due to the exclusive locks) ?

If the answer is no, do you have a plan for this kind of situation ( but in this case the question would be appropriate in postgres community ).

Thank you for your time 

Best regards

**Comment by revolunet** [comment]
You need to provide your image that have the extension and create the extension with `postInitTemplateSQL `

more : https://cloudnative-pg.io/documentation/1.18/postgis/#create-a-new-postgresql-cluster-with-postgis

**Comment by thomasboussekey** [comment]
Hello,

Having the same need, I have created [this PR](https://github.com/cloudnative-pg/postgres-containers/pull/74), in order to add `pg_repack` eextension into cloudnative-pg default containers.

**Comment by gbartolini** [comment]
While I understand the desire to have extensions in the default containers, we also need to consider the default containers as potential base images for further customizations. I think that as a community, we should define some guidelines to include extensions to the default images (and I now feel guilty by having requested and merged pgvector extension - which however is not invasive and only operates on the user space).

We also need to consider the level of supportability from the community on the distributed images. I am open to ideas, and I think we should start working on such a document collectively. Another proposal could be to have another repository in this project containing a "larger" version of the images, managed by the broader community, that is endorsed by CloudNativePG community and that are built starting from the base images.

However, I really think it is time to ratify this process and write some guidelines around it. I am more than happy to discuss removing pgvector from the default image if needed. Thoughts?

**Comment by thomasboussekey** [comment]
Hello @gbartolini ,

Thank you for the feedback!

Our migration project is close to go live. One of the last blocking point for us is the ability to remove BLOAT for our user tables.
On our current stack, we are using `pg_repack` to help us on this task.

Maybe, we have to change our habits in order to remove BLOAT, and I'm  open to new techniques, feel free to provide tips on that way.

It is very cool to provide 2 default container images (with or without `postgis`).
If the 2 default images needs to **stay minimistic** (smallest size and a very few extensions), is there a possibility to add new container images with other extensions (pg_repack, pg_vector ... :warning: of cartesian product of possibilities) or just a 3rd one in `all-inclusive` mode :wink: (containing all extensions).

> **note:** the larger container images could be built at a lower frequency

What do you think of this approach?
If you are OK, I can try to code a possible implementation.

I don't know `pgvector`, and I'm not jealous of the fact that it has been added on  container images :wink:

**Comment by sxd** [comment]
Hi @thomasboussekey 

First thing to keep in mind, having a lot of extensions inside an image it's not a good approach since it will increase the size of the image for people that only requires PostgreSQL, the idea of the images we provide it's to let people create their own custom images with the extensions they need.

Currently we added a couple of extensions, pgvector was the last one, and this was a good example on how to add more than one extension so people can start creating their own images, but that's it, adding more extensions means more work and means that people will ask us to support those extensions. What do we mean by support those extensions? well, we use these images to for the E2E tests, so, if for some reason one of those extensions affects the native PostgreSQL behavior that will crash our tests and will take us lot of time to know that it was one extension and even more time find which extension was, and that will take the focus out off the operator deployment, and that's my personal opinion.

Also, what's stopping you of creating and maintain your own image? is there's something we can do to help with that? Was the idea of having internal images not considered inside your project?

With that being said, we also need help to maintain the images and keep them up to day and there's no more extensions! so if we want to provide a bigger set of images, for sure we will require more help.

Let's open the discussion about the extensions and not focus only on one extension.

Regards,

**Comment by thomasboussekey** [comment]
Hello @sxd & @gbartolini ,

I agree with your point of view, keeping things as simple as possible is  important.

I will be very happy to keep your `postgis` images for my containers.
It would be easier for me not to have to build an image factory.

Can you provide me a **workaround** to the usage of the extension `pg_repack`, in order to remove BLOAT on my heavily modified user tables?
`REINDEX` & `VACUUM` is not  as efficient as repacking :disappointed: 


Thanks in advance,
Have a great week-end,
Thomas

**Comment by davidpechcz** [comment]
Just to add my two cents to the discussion - I've never seen any production installation of Postgres without `pg_repack` for many years. To me, it's on the same level `pg_stat_statements` and I strongly believe that it should be included by default. It's just not "any extension", but without any stats, I believe it's used as a de facto standard replacing problems of basic PG operations and once you have any larger tables, you need it.

However I think that it's not preferred to `kubectl exec` inside Pods, so "ideal" would be also some support like `kubectl cnpg X` here.
